elastic/kibana
# [EPM] Update UI to handle package versions and updates

https://github.com/elastic/kibana/issues/64255

- Does not check for whether the user has auto updates turned on or make any mention of it because auto updating functionality doesn't exist yet and unlikely will for Alpha
- Update the Settings tab for package detail page to display version information and update option
- When the user clicks on a package from the list page they will go to the version they have installed
- Fixes a bug where delete promises of saved objects assets were not being returned and awaited and resolving after some assets were installed so that the next time a delete or reinstall occurred, they did not exist

## Testing instructions:
This scenario where user's have out of date packages in the UI would happen if the user had auto updates turned off which will be the case for alpha but the plan is to make auto updates default.   

1. - Install an out of date package through the API: 
POST http://localhost:5603/ssg/api/ingest_manager/epm/packages/nginx-0.0.1
2. - Go to the package list page and click on the "installed integrations" tab.  
3. - Click on the installed Nginx package and notice you are redirected to the version installed (0.0.1) and not the latest (0.0.2)
4. - Try to update the package

User has the most up to date version of package installed
<img width="1140" alt="Screen Shot 2020-04-28 at 2 19 09 PM" src="https://user-images.githubusercontent.com/1676003/80522898-3efcb700-895b-11ea-9db5-a1d5709c0bc5.png">

User has out of date version installed
<img width="1179" alt="Screen Shot 2020-04-28 at 2 22 01 PM" src="https://user-images.githubusercontent.com/1676003/80523162-a4e93e80-895b-11ea-8f09-a31ee3f44e82.png">

User clicks the "update" button.  Uninstall button is hidden so they don't try to uninstall during an update
<img width="1343" alt="Screen Shot 2020-04-28 at 2 22 42 PM" src="https://user-images.githubusercontent.com/1676003/80523276-cba77500-895b-11ea-9acb-d378aa03653c.png">

A successful update.  they are redirected to the new package version page (change route) and there is a success toast
<img width="1132" alt="Screen Shot 2020-04-28 at 2 23 54 PM" src="https://user-images.githubusercontent.com/1676003/80523347-ec6fca80-895b-11ea-9289-e16394bb2396.png">

An unsuccessful update attempt
<img width="1417" alt="Screen Shot 2020-04-28 at 2 42 41 PM" src="https://user-images.githubusercontent.com/1676003/80524998-905a7580-895e-11ea-80c6-39d2a2a47346.png">

#### Edgecases where the user navigates to a version that is not the one they have installed (perhaps from a bookmark or history as we always link to the version they have installed or the most recent if not installed).  

User goes to package version that is newer than the one they have installed.  I am not too concerned about this view because they would have had to guess this URL as they always get sent to the version they have installed if they have one installed and update from that version
Navigate to http://localhost:5601/ssg/app/ingestManager#/epm/detail/nginx-0.0.2 with nginx-0.0.1 installed
<img width="1136" alt="Screen Shot 2020-04-28 at 2 49 30 PM" src="https://user-images.githubusercontent.com/1676003/80525622-88e79c00-895f-11ea-98f6-e6d3fd7e4b8b.png">

Navigate to http://localhost:5601/ssg/app/ingestManager#/epm/detail/nginx-0.0.1 with nginx-0.0.2 installed
User directly navigates to the url of a package that is older than the version they have installed.  There are no options for installing or removing the package because its not the version they have installed.
<img width="1127" alt="Screen Shot 2020-04-28 at 2 56 25 PM" src="https://user-images.githubusercontent.com/1676003/80526246-76219700-8960-11ea-9319-9a837d94bc19.png">

For the edgecases, Henry and I discussed possibility of changing the package detail route from  `/detail/nginx-0.0.1/` to `/detail/nginx` to simplify things, but this would be an issue for when we support installing multiple versions at once

@ruflin should we be blocking installations or updates to a version that is not the latest version?
